Main.hs

```diff
@@ -3,27 +3,87 @@ module Main where
 import Control.Monad (unless)
 import Text.Printf (printf)
 
-data Expr
+data Operator = Add 
+     | Sub 
+     | Mult 
+     | Div 
+     | Pow
+     deriving (Eq)
+     
+instance Show Operator where
+     show Add = "+"
+     show Sub = "-"
+     show Mult = "*"
+     show Div = "/"
+     show Pow = "^"
 
-instance Show Expr where
-  show = undefined
+data Expr = Num Double 
+     | Sqrt Expr 
+     | CompExpr Operator Expr Expr
 
+instance Show Expr where
+    show (Num a) = show a
+    show (Sqrt a) = printf "sqrt(%s)" (show a)
+    show (CompExpr op a b) = printf "(%s %s %s)" (show a) (show op) (show b)
+    
 instance Eq Expr where
-  (==) = undefined
+  Num a == Num b = a == b
+  Sqrt a == Sqrt b = a == b
+  CompExpr op a b == CompExpr op' a' b' = op == op' && a == a' && b == b
+  _ == _ = False
```

Comment:
Whenever you have the default implementation of a derivable type class, it's better to derive it.